### PR TITLE
Crash when displaying shallow nested recent files saved on network drive

### PR DESCRIPTION
**Pull request by holyjewsus**
### Purpose

This PR changes the file path converter for displaying recent file paths in the recent file submenu in the file menu. Because network drives or servers file paths include extra '\'s and don't return roots the same as regular file paths this code broke on shallowly nested paths. (< 3 directories) 
 Instead of querying for directory.parent and parent.parent we loop some threshold and return when we encounter a null parent.

I have set this threshold to 2 directories under the root directory, this seems to be similar the original intent of filtering by 5 slashes. (which are escaped to `\\`)

http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7929

screen shot 2015-08-11 at 4 02 58 pm

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] The level of testing this PR includes is appropriate

### Reviewers

@lukechurch 

### FYIs

@pboyer - just want to make sure I interpreted the intent of the original code correctly.

### CODE REVIEW

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 1]
```diff
@@ -1581,20 +1581,36 @@ public object Convert(object value, Type targetType, object parameter, CultureIn
...
+                
+                var currentDirInfo = new DirectoryInfo(Path.GetDirectoryName(str));
+                var root = currentDirInfo.Root;
+                var rootName = root.FullName;
+
+                var collapsed = new List<string>();
+                collapsed.Add(name);
+                
+                var count = 0;
+                while(count < threshold)
```
Comment (lukechurch) [inline src/DynamoCoreWpf/UI/Converters.cs]: Prefer for loops to while loops. They're more concise, easier to read and harder to make some kinds of mistakes with.

for (int count = 0; count < threshold; count++) {

...
}

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 2]
```diff
@@ -1581,20 +1581,36 @@ public object Convert(object value, Type targetType, object parameter, CultureIn
             // if the number of directories deep exceeds threshold
             if (str.Length - str.Replace(@"\", "").Length >= 5)
             {
-                var root = Path.GetPathRoot(str);
+                //directories to go down under the root
+                var threshold = 2;
```
Comment (lukechurch) [inline src/DynamoCoreWpf/UI/Converters.cs]: threshold for what? Consider a more specific name MAX_NUMBER_OF_FOLDERS = 2

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 3]
```diff
@@ -1581,20 +1581,36 @@ public object Convert(object value, Type targetType, object parameter, CultureIn
             // if the number of directories deep exceeds threshold
             if (str.Length - str.Replace(@"\", "").Length >= 5)
             {
```
Comment (lukechurch) [inline src/DynamoCoreWpf/UI/Converters.cs]: This is marginally too complicated for its own good :)

Given that we've seen this crash, I think we would be justified in extracting this into a helper method and writing a couple of trivial regression unit tests.

**Comment by lukechurch** [comment]
couple of comments Mike. Hope this doesn't feel completely over the top.

**Comment by holyjewsus** [comment]
@lukechurch PTAL

**Comment by holyjewsus** [comment]
@lukechurch it appears that the eng ops build completed without errors but updating through the githubAPI failed so I would consider that yellow light green...

File: test/DynamoCoreWpfTests/ConverterTests.cs [hunk 4]
```diff
@@ -593,5 +593,28 @@ public void LibraryViewModeToBoolConverterTest()
             result = converter.Convert(SearchViewModel.ViewMode.LibrarySearchView, null, null, null);
             Assert.IsFalse((bool)result);
         }
+
+        [Test]
+        public void FilePathDisplayConverterTest()
+        {
+            var converter = new FilePathDisplayConverter();
+
+            Assert.DoesNotThrow(() => converter.Convert("\\\\psf\\Home\\Desktop\\somedyn.dyn",null,null,null));
```
Comment (lukechurch) [inline test/DynamoCoreWpfTests/ConverterTests.cs]: Stylistic nit - I'd before @"..." here, to stop this crazy Windows \\\\\\\\\\\\\\\\\ing

**Comment by lukechurch** [comment]
lgtm, thanks @holyjewsus

**Pull request merged**